### PR TITLE
Atree Inlining & Deduplication with Cadence v0.42

### PR DESCRIPTION
**Pull request by fxamacker**
This PR is to be used for building migration program needed by spork and running benchmarks.

This enables Atree Inlining & Deduplication with Cadence 0.42 in flow-go (e.g. atree migration program, etc.).

- Epic https://github.com/onflow/atree/issues/292
- Cadence 0.42 Integration https://github.com/onflow/cadence/pull/2882

PR #5349 (in fxamacker branch) will be closed and we will start using this PR in a feature branch to build migration program.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -50,8 +50,8 @@ require (
 	github.com/multiformats/go-multiaddr v0.12.2
 	github.com/multiformats/go-multiaddr-dns v0.3.1
 	github.com/multiformats/go-multihash v0.2.3
-	github.com/onflow/atree v0.6.0
-	github.com/onflow/cadence v0.42.10
+	github.com/onflow/atree v0.6.1-0.20240405171204-c6c951de7189
+	github.com/onflow/cadence v0.42.10-0.20240409201024-d000241476d6
```
Comment (turbolent) [inline go.mod:54]: I tagged `v0.42.11-atree-register-inlining` in the Cadence repo, could you please update to that one? 
It is `0.42.10` + atree register inlining (.11 to make the version "newer" than v0.42.10)

**Review by turbolent — APPROVED** [review]
👍 

Just one comment regarding the Cadence version. We can also do that in another PR if you want

File: go.mod (hunk 1 above)
Comment (fxamacker) [reply go.mod]: Thanks for tagging!

Updated this PR to use it in commit [5c24862](https://github.com/onflow/flow-go/pull/5736/commits/5c24862d69bc0ef2c22c5175ae0438f82677210a).

File: fvm/evm/emulator/state/collection.go [hunk 2]
```diff
@@ -47,16 +47,16 @@ func NewCollectionProvider(
...
 		return nil, err
 	}
-	// sanity check the storage ID address
-	if storageID.Address != cp.rootAddr {
-		return nil, fmt.Errorf("root address mismatch %x != %x", storageID.Address, cp.rootAddr)
-	}
+	// TODO: sanity check the slab ID address
+	// if slabID.Address != cp.rootAddr {
+	// return nil, fmt.Errorf("root address mismatch %x != %x", storageID.Address, cp.rootAddr)
+	// }
```
Comment (turbolent) [inline fvm/evm/emulator/state/collection.go:57]: Is this a TODO for this PR?

**Review by turbolent — COMMENTED** [review]
LGTM, just one question re: the TODO

File: fvm/evm/emulator/state/collection.go (hunk 2 above)
Comment (fxamacker) [reply fvm/evm/emulator/state/collection.go]: Good catch! :+1:  Added check back in commit [62c7780](https://github.com/onflow/flow-go/pull/5736/commits/62c77802b05e0dc93a3139c1bde3ed2461d4354b).

**Review by turbolent — APPROVED** [review]

**Comment by turbolent** [comment]
CI seems to fail because of flaky tests, can probably ignore those

**Comment by turbolent** [comment]
I think the branch protection rules are broken again and so the CI check list requires checks that no longer exist/got renamed. Merging

**Pull request merged**